ts/src/main/scala/ch/epfl/ts/component/ComponentBuilder.scala

```diff
@@ -0,0 +1,157 @@
+  var instances = List[ComponentRef]()
+
+  /** Responsible for sending `PoisonPill`s when we want to terminate the system gracefully */
+  private val reaper = system.actorOf(Props(classOf[Reaper]), "Reaper")
+
+  /**
+   * "Builder" actors that are empty but useful to create instances
+   * under their name. This way, we create automatically an easy-to-query
+   * actor hierarchy based on concrete Actor type.
+   */
+  private lazy val roots: Seq[(Class[_], ActorRef)] = {
+    val rootNames = Seq(
+      classOf[Trader] -> "traders",
+      classOf[Evaluator] -> "evaluators",
+      classOf[Broker] -> "brokers",
+      classOf[FetchingComponent] -> "fetchers",
+      classOf[MarketSimulator] -> "markets",
+      classOf[Timekeeper] -> "timekeepers",
+      classOf[Any] -> "other"
+    )
+    
+    rootNames.map({
+      case (clazz, name) => clazz -> system.actorOf(Props(classOf[ParentActor]), name)
+    })
+  }
+  
+  def getRootForClass(clazz: Class[_]): ActorRef = {
+    val opt = roots.find(p => p._1.isAssignableFrom(clazz))
+  	// Should not fail since we have an `Any` entry in `roots`
+    opt.get._2
+  }
+  
+  
+  /**
+   * Connect `src` to `dest` on the given type of messages
+   */
+  def add(src: ComponentRef, dest: ComponentRef, data: Class[_]) {
+    println("Connecting " + src.ar + " to " + dest.ar + " for type " + data.getSimpleName)
+    graph = graph + (src -> ((dest, data) :: graph.getOrElse(src, List[(ComponentRef, Class[_])]())))
+    src.ar ! ComponentRegistration(dest.ar, data, dest.name)
+  }
+
+  def add(src: ComponentRef, dest: ComponentRef) = (src, dest, classOf[Any])
+
+  def start = instances.map(cr => {
+    cr.ar ! StartSignal
+    println("Sending start Signal to " + cr.ar)
+  })
+
+  /**
+   * Send a `StopSignal` to all managed components, giving them the opportunity
+   * to run some cleanup.
+   */
+  def stop = instances.map(cr => {
+    cr.ar ! StopSignal
+    println("Sending stop Signal to " + cr.ar)
+  })
+
+  /**
+   * Create a new component with the given name using the `props`.
+   * Certain types of components such as `Trader`s, `Evaluator`s, `Market`s, etc
+   * are automatically created under an empty root actor corresponding to their class.
+   * This allows us to query the actor hierarchy easily for given types.
+   * 
+   * @warning Current implementation is blocking
+   * @TODO Any way to make this non-blocking?
+   */
+  def createRef(props: ComponentProps, name: String) = blocking {
+    // Determine the root actor to use for this class
+    val root = getRootForClass(props.actorClass())
+
+    // Ask the relevant root actor to make an instance for us
+    implicit val timeout = new Timeout(1 second)
+    val future = (root ? ParentActor.Create(props, name)).mapTo[ParentActor.Done]
+    val ref = Await.result(future, timeout.duration).ref
+    
+    // Wrap it into a ComponentRef
+    instances = new ComponentRef(ref, props.actorClass(), name, this) :: instances
+    instances.head
+  }
```

Comment:
Nice! I like how you integrated the root actors without breaking the external interface